src/internet_identity/src/anchor_management/registration/registration_flow_v2.rs

```diff
@@ -0,0 +1,172 @@
+            captcha_png_base64: captcha_png_base64.0,
+        },
+    })
+}
+
+async fn captcha_flow_state(flow_created_timestamp_ns: u64) -> (Base64, RegistrationFlowState) {
+    let mut rng = &mut make_rng().await;
+    let (captcha_png_base64, captcha_solution) = create_captcha(&mut rng);
+    let flow_state = RegistrationFlowState::CheckCaptcha {
+        flow_created_timestamp_ns,
+        captcha_solution,
+        captcha_png_base64: captcha_png_base64.clone(),
+    };
+    (captcha_png_base64, flow_state)
+}
+
+pub async fn check_captcha(arg: CheckCaptchaArg) -> Result<IdRegNextStepResult, CheckCaptchaError> {
+    let Some(current_state) = state::with_flow_states(|s| s.registration_flow_state(&caller()))
+    else {
+        return Err(CheckCaptchaError::NoRegistrationFlow);
+    };
+
+    let RegistrationFlowState::CheckCaptcha {
+        flow_created_timestamp_ns,
+        captcha_solution,
+        ..
+    } = current_state
+    else {
+        return Err(CheckCaptchaError::UnexpectedCall {
+            next_step: RegistrationFlowNextStep::from(current_state),
+        });
+    };
+
+    let caller = caller();
+    if check_captcha_solution(arg.solution, captcha_solution).is_err() {
+        let (captcha_png_base64, flow_state) = captcha_flow_state(flow_created_timestamp_ns).await;
+        state::with_flow_states_mut(|flow_states| {
+            flow_states.update_registration_flow(caller, flow_state)
+        })
+        // If we fail to update the flow, then the flow has expired in the time it took to create
+        // the captcha.
+        .map_err(|_| CheckCaptchaError::NoRegistrationFlow)?;
+
+        return Err(CheckCaptchaError::WrongSolution {
+            new_captcha_png_base64: captcha_png_base64.0,
+        });
+    }
+
+    let flow_state = RegistrationFlowState::FinishRegistration {
+        flow_created_timestamp_ns,
+    };
+
+    state::with_flow_states_mut(|flow_states| {
+        flow_states.update_registration_flow(caller, flow_state)
+    })
+    // The update_registration_flow triggers a clean-up as well, which could lead to a situation where
+    // the flow being processed here is cleaned up. In that case, abort with an error.
+    .map_err(|_| CheckCaptchaError::NoRegistrationFlow)?;
+
+    Ok(IdRegNextStepResult {
+        next_step: RegistrationFlowNextStep::Finish,
+    })
+}
+
+pub fn identity_registration_finish(
+    arg: IdRegFinishArg,
+) -> Result<IdRegFinishResult, IdRegFinishError> {
+    let Some(current_state) = state::with_flow_states(|s| s.registration_flow_state(&caller()))
+    else {
+        return Err(IdRegFinishError::NoRegistrationFlow);
+    };
+
+    let RegistrationFlowState::FinishRegistration { .. } = current_state else {
+        return Err(IdRegFinishError::UnexpectedCall {
+            next_step: RegistrationFlowNextStep::from(current_state),
+        });
+    };
+
+    let identity_number = create_identity(&arg)?;
+
+    let caller = caller();
```

Comment:
Is there a reason not to create this var the first thing and use it in line 117?

Reply:
Actually no, I'll move it up.